apps/server/src/shared/repo/task/task.repo.integration.spec.ts

```diff
@@ -45,6 +45,230 @@ describe('TaskRepo', () => {
+					users: [student2],
+				});
+
+				const mathsTask2 = taskFactory.build({
+					name: 'all primes under million',
+					course: mathsCourse,
+					users: [student3],
+				});
+
+				const mathsTask3 = taskFactory.build({
+					name: 'algebra 1',
+					lesson: algebraLesson,
+					users: [student2],
+				});
+
+				await em.persistAndFlush([
+					teacher1,
+					teacher2,
+					student1,
+					student2,
+					student3,
+					student4,
+					englishCourse,
+					grammerLesson,
+					historyCourse,
+					mathsCourse,
+					algebraLesson,
+					englishTask1,
+					englishTask2,
+					englishTask3,
+					englishTask4,
+					historyTask1,
+					historyTask2,
+					historyTask3,
+					mathsTask1,
+					mathsTask2,
+					mathsTask3,
+				]);
+				em.clear();
+
+				return {
+					teacher1,
+					teacher2,
+					student1,
+					student2,
+					student3,
+					student4,
+					englishCourse,
+					grammerLesson,
+					historyCourse,
+					mathsCourse,
+					algebraLesson,
+					englishTask1,
+					englishTask2,
+					englishTask3,
+					englishTask4,
+					historyTask1,
+					historyTask2,
+					historyTask3,
+					mathsTask1,
+					mathsTask2,
+					mathsTask3,
+				};
+			};
+
+			it('Assigned users should not affect all courses', async () => {
+				const { englishCourse, mathsCourse, historyCourse } = await setup();
+				const [, total] = await repo.findAllByParentIds({
+					courseIds: [englishCourse.id, mathsCourse.id, historyCourse.id],
+				});
+				expect(total).toBe(6);
+			});
+
+			it('Filter by assigned users and courses', async () => {
+				const { englishCourse, mathsCourse, historyCourse, student1 } = await setup();
+				const [, total] = await repo.findAllByParentIds(
+					{
+						courseIds: [englishCourse.id, mathsCourse.id, historyCourse.id],
+					},
+					{
+						userId: student1.id,
+					}
+				);
+				expect(total).toBe(3);
+			});
+
+			it('Filter by assigned users and lessons in same course', async () => {
+				const { grammerLesson, algebraLesson, student1 } = await setup();
+				const [result, total] = await repo.findAllByParentIds(
+					{
+						lessonIds: [grammerLesson.id, algebraLesson.id],
+					},
+					{
+						userId: student1.id,
+					}
+				);
+				expect(total).toBe(2);
+				expect(result.map((r) => r.name)).toContain('grammer1');
+				expect(result.map((r) => r.name)).toContain('grammer3');
+			});
+
+			it('Filter by assigned users and lessons in different courses', async () => {
+				const { grammerLesson, algebraLesson, student2 } = await setup();
+
+				const [result, total] = await repo.findAllByParentIds(
+					{
+						lessonIds: [grammerLesson.id, algebraLesson.id],
+					},
+					{
+						userId: student2.id,
+					}
+				);
+				expect(total).toBe(2);
+				expect(result.map((r) => r.name)).toContain('algebra 1');
+				expect(result.map((r) => r.name)).toContain('grammer3');
+			});
+
+			it('Filter by assigned users and lessons in different courses but with course fallback', async () => {
```

Comment:
the test setup is really hard to see through. took some time to me to understand which course the student3 is and with that which lesson is linked to that course to see which tasks are found, as the variable `englishTask4` does not equal the name `grammer3`

Reply:
I think some of these tests are not redundant, as we cover cases from api e2e tests